XPSS rework into sequential difficulty

marsh1832 opened the pull request:
Contract is sequential, starts at 350 m/s and increases by 50 until 800 m/s where it increases by 100 m/s to a maximum of 1500. The speed window is the same as the increment. Pays 2x as much per m/s as compared to the repeatable. This is so that it pays out about the same as if you grinded about 20 XPSS contracts, enough sequential contracts to reach the upper limit in speed that XPSS would offer. The anti-grind bonus has been reduced by about 1/3 due to there being less of an incentive to grind with a sequential contract. Uses the existing data nodes so it should slot in to an existing career. It creates a new value to track completions, so it will start at 350 m/s in an existing career.

Reward might need some balance with aviation engine unlock costs (or the other way around.) Build time of a plane in the orbital era is quite short so the payout is probably well-tuned with VAB time. I've moved the ramjet to the 1970 aviation node. It kind of breaks the contract progression where it is, and it needs a balance pass anyway. Seems to be placed based on the AQM-60. I intend to balance the ramjet based on the D-21 drone that flew missions between 1969-1971. This was an endurance ramjet that is more suitable for a fictional crewed aircraft. It makes sense for this engine to be in the node that follows the J58 both for contract progression and history. I've changed the unlock and cost to be slightly higher than the J58 for now.

I'm still learning github, so I don't know why it's not showing the line-by-line changes in the contract, sorry.

siimav commented:
I think the altitude requirements will also need to increase with speeds. Mach 3+ isn't really feasible below 20km.

marsh1832 commented:
I’ve removed the altitude requirement altogether, in part because this wide speed range will have a wide altitude range as well. But there’s a soft altitude window in a way, between where there’s enough air for combustion and too much drag. I’d rather that be the defining factor of the flight than a specific window in the contract.

siimav commented:
Sorry, I missed that part. For some reason GH says that you've modified the entire file which usually happens if the line endings got switched up.
I think some minimum altitude requirement makes sense though. Especially since the contract uses m/s instead of Mach values. This would also prevent the contract from being completed in long dives.

marsh1832 commented:
Yea I'm not sure what I messed up, is there a way to fix the line endings?

I have some ideas on altitude. Setting just a minimum altitude that increases isn't too hard. What's the inherent problem with completing a speed record in a dive? Isn't that how they were done? Why does it have to be level flight? The soft altitude window I mentioned can be more strictly enforced with a longer time requirement, say 5 minutes rather than 3. Even just 3 minutes is a long time to be diving. A passenger jet in a standard descent profile will lose about 3km of altitude in 3 minutes. So a longer time window can make any significant dives impossible. But does it matter to make them impossible? There's other factors like the engine's temp limits, the airframe temp limits, ect, to consider.

ec429 commented:
I thought the point of the altitude requirement was so that you couldn't complete it with a deathstick ballistic arc at high altitude.  I.e. force it to actually be an aeroplane, under thrust and aerodynamically supported.

marsh1832 commented:
Sounds about right. The contract has a must be jet/no rockets requirement now, specifically it checks for ModuleEnginesRF. I recognize the contract can be cheesed by rocket launching a winged craft first and then ejecting the rocket equipment and managing to fly it in that narrow speed window at a high enough altitude to meet the time requirement, but at that point the player is doing more harm than good to themselves since the profitability of a plane is that you can reuse it, especially considering how KCT edit works. A simple roadblock is a broad upper limit... say 40km. Although the most direct solution is to just increase the time requirement of the flight. Eventually all glider cheese will run out of altitude and required the intended air breathing powered flight.

marsh1832 commented:
I added a +/- 10 m/s rate of climb limitation to the contract to limit glider cheese and dives. This allows 1.8 km of dive, which might allow a marginal plane to complete a contract, but not the next one. In my test I had a plane that did 310 in level flight, and with -10 m/s it got about 330.

Maintaining 10 is trivial with MJ or AA, and fairly easy with MAF if you have a steady hand. Keyboard pilots will have a hard time if they are not familiar with precision control or trim. I added lots of "jet" and "level flight" statements throughout the contract texts for consistency and to maximize the chance the player will see it. I also added an empty line at the bottom which seemed to have fix the reporting of line-by-line changes.

siimav commented:
You can thank @al2me6 for the conflicts he just created 😀
Also I think it would be better if you changed the soft cooldown formula to the one used on the repeatable commsat contracts. Would be easier to manage if there's fewer formulas going around all with slightly different parameters.

marsh1832 commented:
The commsat formula has a similar effect that I was going for so I don't see any issues with balance to use it here as well. Made a grammar change, I think the only error that carried over in this version.

siimav commented:
You also need to merge master into your branch or rebase on master to fix the conflicts. Otherwise the PR cannot be merged.

al2me6 commented:
Ugh, sorry. I did not realize this was here and would cause a merge conflict.

marsh1832 commented:
Should be good now? I'm slowly learning how all this works T.T

siimav commented:
Yep, it's good as far as I can tell. Hopefully you didn't overwrite anything important from @al2me6. Also it appears that whatever you did fixed the line endings as well. 🎉

siimav commented:
<details>
  <summary>Some stats from the spyware on how much various people have earned from the XPSS contract:</summary>
  
```JS
[
  {
    "sum": 188300,
    "count": 14,
    "careerId": "5f972381485e860ba4648576",
    "careerName": "siimav's non-caveman career",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-05-08T08:34:58.714Z"
  },
  {
    "sum": 126800,
    "count": 5,
    "careerId": "5f972901485e860ba4648577",
    "careerName": "Data stolen from Calvin Maclure",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1952-09-26T13:02:49.637Z"
  },
  {
    "sum": 77200,
    "count": 6,
    "careerId": "5fa6fb783f449cc4a7b8574d",
    "careerName": "lpg's career #13",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-02-12T13:50:45.778Z"
  },
  {
    "sum": 138000,
    "count": 9,
    "careerId": "5fa89363de28d0ff720376de",
    "careerName": "The Destroyers' 2020 run",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-03-24T11:55:13.186Z"
  },
  {
    "sum": 231300,
    "count": 20,
    "careerId": "6006de89a376bdea131cd1f1",
    "careerName": "Butcher down under",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-05-24T00:45:35.861Z"
  },
  {
    "sum": 60880,
    "count": 9,
    "careerId": "60424449823ebb84e3e0cad5",
    "careerName": "Leudaimon_Hard_SR+LunarNerf",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-04-09T20:38:04.306Z"
  },
  {
    "sum": 72100,
    "count": 3,
    "careerId": "607867e29b33b605a63b333b",
    "careerName": "Pap-USA",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-11-18T12:17:55.973Z"
  },
  {
    "sum": 71700,
    "count": 3,
    "careerId": "60858c17331df764e1a235db",
    "careerName": "theysen's rusty moderate approach",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1953-03-26T12:40:20.179Z"
  },
  {
    "sum": 129195,
    "count": 7,
    "careerId": "60912337436d72349b5982d7",
    "careerName": "happz's RP-1 career #1",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1952-11-07T15:27:54.498Z"
  },
  {
    "sum": 179500,
    "count": 7,
    "careerId": "6091c417436d72349b5982da",
    "careerName": "Kobymaru-v1_8",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1952-12-31T07:30:55.805Z"
  },
  {
    "sum": 68300,
    "count": 2,
    "careerId": "60990b0a436d72349b5982e2",
    "careerName": "NathanKell's Rusty",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1958-12-01T14:22:53.116Z"
  },
  {
    "sum": 34000,
    "count": 3,
    "careerId": "60a180b54138ff20068ad0a1",
    "careerName": "Ernos' Career",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1955-06-11T17:48:09.895Z"
  },
  {
    "sum": 46080,
    "count": 1,
    "careerId": "60af987f546635b6297c8cfa",
    "careerName": "Charon Unlimited (bongotastic)",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1954-05-22T23:38:08.764Z"
  },
  {
    "sum": 65400,
    "count": 3,
    "careerId": "60c0d7d4546635b6297c8cff",
    "careerName": "daTodi's (mainly) soviet career #1",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1952-08-13T01:11:06.086Z"
  },
  {
    "sum": 56000,
    "count": 2,
    "careerId": "60c22b5e546635b6297c8d01",
    "careerName": "Al2Me6's First Career",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1953-02-18T12:56:41.935Z"
  },
  {
    "sum": 198800,
    "count": 13,
    "careerId": "60d3b8c9546635b6297c8d14",
    "careerName": "Elouda ModerateUchinoura",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-06-12T00:31:23.448Z"
  },
  {
    "sum": 30400,
    "count": 1,
    "careerId": "60d6d3ca546635b6297c8d16",
    "careerName": "Grants historic-ish playthrough",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1953-08-24T04:16:05.171Z"
  },
  {
    "sum": 77200,
    "count": 2,
    "careerId": "60d9f4b84352e9ff31c97bdd",
    "careerName": "New Age",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1953-01-04T13:53:31.138Z"
  },
  {
    "sum": 87600,
    "count": 8,
    "careerId": "60de06f0e82eb7d3c8b1c7d8",
    "careerName": "Aazards Caveman Career to Point of Ban",
    "contractInternalName": "CrewedSupersonic",
    "contractDisplayName": "X-Planes (Supersonic)",
    "type": 1,
    "date": "1951-02-04T21:52:45.213Z"
  }
]
```
</details>

marsh1832 commented:
The way I balanced the rewards was comparing the current XPSS total rewards against the rewards for completing the rework 8 times, which is up to and including the 700 m/s, since that's the upper limit of the previous contract and the tech which would complete it.

Total rewards for the rework up to 700 m/s would be 137,600. I'm surprised some of the grinders got even more than that, but this is about what I was aiming for. Somewhere in the middle, where most people will either not have to grind as many to get the same reward, or they will get more reward for what they did do.

What tier of grinder should the rewards be based around? I could see it being lowered to ~80k for the less grindy player, or up to maybe 200k to cover the most grindy.

The post 700 m/s contracts, up to the final one of 1500 m/s, pay an additional 280,200 in total.

siimav commented:
I feel like 42600 from the first 3 that the J47 can complete could be a bit low. Also what's the thermal limit on J57?

marsh1832 commented:
It’s been a while since I’ve tested it, but I think it’s in the 700’s. Also if I recall cockpit heating starts to be a real problem too, can’t really get away with slowly increasing speed, have to get to it quickly, and the x-15 cockpit will be required for heat tolerance at some point.

siimav commented:
I guess leave the rewards as-is for now. Otherwise we would need to revise the XPD and XPH rewards as well since XPSS would pay significantly better than those.
